sdk/helper/random/string_generator.go

```diff
@@ -0,0 +1,311 @@
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"time"
+	"unicode"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+var (
+	LowercaseCharset   = sortCharset("abcdefghijklmnopqrstuvwxyz")
+	UppercaseCharset   = sortCharset("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	NumericCharset     = sortCharset("0123456789")
+	FullSymbolCharset  = sortCharset("!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~")
+	ShortSymbolCharset = sortCharset("-")
+
+	AlphabeticCharset              = sortCharset(UppercaseCharset + LowercaseCharset)
+	AlphaNumericCharset            = sortCharset(AlphabeticCharset + NumericCharset)
+	AlphaNumericShortSymbolCharset = sortCharset(AlphaNumericCharset + ShortSymbolCharset)
+	AlphaNumericFullSymbolCharset  = sortCharset(AlphaNumericCharset + FullSymbolCharset)
+
+	LowercaseRuneset   = []rune(LowercaseCharset)
+	UppercaseRuneset   = []rune(UppercaseCharset)
+	NumericRuneset     = []rune(NumericCharset)
+	FullSymbolRuneset  = []rune(FullSymbolCharset)
+	ShortSymbolRuneset = []rune(ShortSymbolCharset)
+
+	AlphabeticRuneset              = []rune(AlphabeticCharset)
+	AlphaNumericRuneset            = []rune(AlphaNumericCharset)
+	AlphaNumericShortSymbolRuneset = []rune(AlphaNumericShortSymbolCharset)
+	AlphaNumericFullSymbolRuneset  = []rune(AlphaNumericFullSymbolCharset)
+
+	// DefaultStringGenerator has reasonable default rules for generating strings
+	DefaultStringGenerator = StringGenerator{
+		Length: 20,
+		Rules: []Rule{
+			Charset{
+				Charset:  LowercaseRuneset,
+				MinChars: 1,
+			},
+			Charset{
+				Charset:  UppercaseRuneset,
+				MinChars: 1,
+			},
+			Charset{
+				Charset:  NumericRuneset,
+				MinChars: 1,
+			},
+			Charset{
+				Charset:  ShortSymbolRuneset,
+				MinChars: 1,
+			},
+		},
+	}
+)
+
+func sortCharset(chars string) string {
+	r := runes(chars)
+	sort.Sort(r)
+	return string(r)
+}
+
+// Rule to assert on string values.
+type Rule interface {
+	// Pass should return true if the provided value passes any assertions this Rule is making.
+	Pass(value []rune) bool
+
+	// Type returns the name of the rule as associated in the registry
+	Type() string
+}
+
+// StringGenerator generats random strings from the provided charset & adhering to a set of rules. The set of rules
+// are things like Charset which requires a certain number of characters from a sub-charset.
+type StringGenerator struct {
+	// Length of the string to generate.
+	Length int `mapstructure:"length" json:"length"`
+
+	// Rules the generated strings must adhere to.
+	Rules serializableRules `mapstructure:"-" json:"rule"` // This is "rule" in JSON so it matches the HCL property type
+
+	// Charset to choose runes from. This is computed from the rules, not directly configurable
+	charset runes
+
+	// rng for testing purposes to ensure error handling from the crypto/rand package is working properly.
+	rng io.Reader
+}
+
+// Generate a random string from the charset and adhering to the provided rules.
+func (g *StringGenerator) Generate(ctx context.Context) (str string, err error) {
```

Comment:
I think we should pass the random reader object as a parameter here so on enterprise we can use an external entropy source as the reader

Reply:
Since I already have the reader as a part of the struct (originally for testing, but I realized it would probably be useful for this scenario) does it make more sense to make the reader exposed instead of passing it to the generate function? If we keep it as a part of the struct, we could add the reader to the parse functions.

Reply:
I think it makes the most sense as a passed parameter. That way, in the future, we can cache password policy objects but determine which random source to use based off of the caller context. For instance if we have two database mounts using the password policy, but only one of the mounts is configured to use an external entropy source.

Reply:
:+1: